include/aarch64/kasan.h

```diff
@@ -0,0 +1,36 @@
+#ifndef _AARCH64_KASAN_H_
+#define _AARCH64_KASAN_H_
+
+#include <aarch64/vm_param.h>
+
+/* Part of internal compiler interface */
+#define KASAN_SHADOW_SCALE_SHIFT 3
+
+/* Shadow memory */
+#define KASAN_MD_SHADOW_START 0xffffff0000000000
+#define KASAN_MD_SHADOW_SIZE (1ULL << 30) /* 1 GB */
+static_assert(KASAN_MD_SHADOW_SIZE % SUPERPAGESIZE == 0,
+              "Shadow memory is described by several full PDEs");
+#define KASAN_MD_SHADOW_END (KASAN_MD_SHADOW_START + KASAN_MD_SHADOW_SIZE)
+
+/* Sanitized memory (accesses within this range are checked) */
+#define KASAN_MD_SANITIZED_START KERNEL_SPACE_BEGIN
+#define KASAN_MD_SANITIZED_SIZE                                                \
+  (KASAN_MD_SHADOW_SIZE << KASAN_SHADOW_SCALE_SHIFT)
+#define KASAN_MD_SANITIZED_END                                                 \
+  (KASAN_MD_SANITIZED_START + KASAN_MD_SANITIZED_SIZE)
+
+/* Note: this offset has also to be explicitly set in CFLAGS_KASAN */
+#define KASAN_MD_OFFSET                                                        \
+  (KASAN_MD_SHADOW_START -                                                     \
+   (KASAN_MD_SANITIZED_START >> KASAN_SHADOW_SCALE_SHIFT))
+
+__always_inline static inline int8_t *kasan_md_addr_to_shad(uintptr_t addr) {
+  return (int8_t *)(KASAN_MD_OFFSET + (addr >> KASAN_SHADOW_SCALE_SHIFT));
+}
+
+__always_inline static inline bool kasan_md_addr_supported(uintptr_t addr) {
+  return addr >= KASAN_MD_SANITIZED_START && addr < KASAN_MD_SANITIZED_END;
```

Comment:
Yes and no.
We can't sanitize address greater than `_kasan_sanitized_end` but we should't make any access to that addresses. So in case of KASAN error we will get kernel panic.
For me it was useful during writing that code but if you want I can change that for AArch64 and for mips.

Reply:
I don't get it. If we use `KASAN_MD_SANITIZED_END` instead of `_kasan_sanitized_end`, then once `_kasan_sanitized_end` grows beyond `KASAN_MD_SANITIZED_END`, we won't check addresses between the two even though we should.

Reply:
KASAN shouldn't grow beyond `KASAN_MD_SANITIZED_END`.
Seriously, before you reach `KASAN_MD_SANITIZED_END` you should be out of physical memory.
And it's useful to sanitize access to memory with va higher than `_kasan_sanitized_end` because it's invalid memory. If you modify `boot.c` in wrong way and set `_kasan_sanitized_end` to incorrect value you still want to get an error "invalid reference to shadow map".
I pushed change with limit for `kasan_grow`. Is it enough for you?

Reply:
Ok, I see. I was misinterpreting the meaning of `KASAN_MD_SANITIZED_END`: it's the upper limit for the value of `_kasan_sanitized_end`. The name is terrible. Same for `KASAN_MD_SHADOW_SIZE`: It's the __maximum__ shadow size, not the actual shadow size.

Reply:
Maybe `KASAN_MD_MAX_SHADOW_SIZE`?
I'm open to your suggestions.

Reply:
Yes, `KASAN_MD_MAX_SHADOW_SIZE` and `KASAN_MD_MAX_SANITIZED_END` are ok.

Reply:
done